### PR TITLE
Error on duplicate matcher bindings

### LINKED ISSUE #57593: Multiple bindings of the same metavariable in a macro rule aren't detected until matching.
The following macro definition compiles:

```rust
macro_rules! bad {
  ($a:expr, $a:expr) => {}
}
```

Any attempt to match against this macro will result in an error, because the metavariable cannot be bound twice. There is no reason for the compiler to accept this definition as a result, since it always represents a bug, except for backwards compatibility. It should be fixed, either by assuring it won't break the ecosystem or by deprecating the behaviour and denying it the next edition.

### PR DESCRIPTION
**Pull request by mark-i-m**
fix  #57593 

This should not be merged without a crater run and maybe an FCP. Discussion is ongoing at  #57593.

TODO:
- [x] write tests
- [x] crater run
- [x] ~maybe need edition gating?~ not for 1 regression /centril

r? @petrochenkov

### CODE REVIEW

**Comment by petrochenkov** [comment]
Code LGTM, needs green Travis for crater.

**Comment by mark-i-m** [comment]
As with  #57610, I have been getting a seeming unrelated ICE. I tried rebasing, so hopefully this should be passing now.

It's also worthwhile to note that we had breakage in rustc's own test suite, which I fixed in the last commit...

**Comment by mark-i-m** [comment]
The rebase solved the previous issue, but there was another test that needed to be updated.

**Comment by mark-i-m** [comment]
@petrochenkov This looks like its passing locally. Perhaps we can `bors try` now?

**Comment by petrochenkov** [comment]
@craterbot run start=master#e2f221c75932de7a29845c8d6f1f73536ad00c41 end=try#0bfb90d213a93e466b33d078a5b535a2a6f9c5ff mode=check-only

**Comment by mark-i-m** [comment]
The 4 fixed appear to be spurious un-failures. There are two regressions:
- fluid: appears to be spurious
- nes: is caused by these changes

**Comment by petrochenkov** [comment]
@mark-i-m 
Since there are regressions, could you turn the error into a future-compatibility lint?
It's not too much work in this case.

**Comment by Centril** [comment]
@mark-i-m Could you also reach out to the `nes` crate with a PR perhaps?

Seems like we should be able to turn this into a hard error pretty quickly.

**Comment by mark-i-m** [comment]
I opened https://github.com/TrionProg/nes/pull/2

I will turn it into a future compat lint shortly...

**Comment by mark-i-m** [comment]
@petrochenkov Is there a way to get a `NodeId` from a `Span` or `TokenTree`?

**Comment by mark-i-m** [comment]
@petrochenkov @Centril This is now a future incompatibility lint. Should we have an FCP or something?

File: src/test/ui/macros/macro-multiple-matcher-bindings.rs [hunk 1]
```diff
@@ -0,0 +1,25 @@
+// Test that duplicate matcher binding names are caught at declaration time, rather than at macro
+// invocation time.
+//
+// FIXME(mark-i-m): Update this when it becomes a hard error.
+
+// compile-pass
+
+#![allow(unused_macros)]
```
Comment (petrochenkov) [inline src/test/ui/macros/macro-multiple-matcher-bindings.rs]: Hmm, all `unused` warnings should be disabled by default in UI tests, so this shouldn't be necessary.

File: src/test/ui/macros/macro-multiple-matcher-bindings.rs [hunk 2]
```diff
@@ -0,0 +1,25 @@
+// Test that duplicate matcher binding names are caught at declaration time, rather than at macro
+// invocation time.
+//
+// FIXME(mark-i-m): Update this when it becomes a hard error.
+
+// compile-pass
```
Comment (petrochenkov) [inline src/test/ui/macros/macro-multiple-matcher-bindings.rs]: If you add `// compile-pass` then all the `//~ WARNING` annotations become ignored unfortunately (https://github.com/rust-lang/rust/issues/56277).
Could you use `deny(duplicate_matcher_binding_name)` to get errors instead?

File: src/librustc/lint/builtin.rs [hunk 3]
```diff
@@ -346,6 +346,12 @@ declare_lint! {
     "outlives requirements can be inferred"
 }
 
+declare_lint! {
+    pub DUPLICATE_MATCHER_BINDING_NAME,
```
Comment (petrochenkov) [inline src/librustc/lint/builtin.rs]: Nit: [Lint naming conventions](https://github.com/rust-lang/rfcs/blob/master/text/0344-conventions-galore.md#lints).
`duplicate_matcher_binding_name` -> `duplicate_matcher_binding_names`
(Not too important for soon-to-be-removed deprecation warnings but still.)

File: src/librustc_lint/lib.rs [hunk 4]
```diff
@@ -341,6 +341,11 @@ pub fn register_builtins(store: &mut lint::LintStore, sess: Option<&Session>) {
             reference: "issue #57571 <https://github.com/rust-lang/rust/issues/57571>",
             edition: None,
         },
+        FutureIncompatibleInfo {
+            id: LintId::of(DUPLICATE_MATCHER_BINDING_NAME),
+            reference: "issue #57593 <https://github.com/rust-lang/rust/issues/57593>",
```
Comment (Centril) [inline src/librustc_lint/lib.rs]: Change this to a new issue with the usual format (see https://github.com/rust-lang/rust/issues/57644 for an example).

**Comment by Centril** [comment]
> @petrochenkov @Centril This is now a future incompatibility lint. Should we have an FCP or something?

Sure, let's do it...

I propose that we introduce this [**C-future-compatibility**](https://github.com/rust-lang/rust/labels/C-future-compatibility) warning ([*1 regression on crater*](https://github.com/rust-lang/rust/pull/57617#issuecomment-455341547)) that catches *multiple macro metavariables with the same name*, e.g. `macro_rules! b { ($a:expr, $a:expr) => {} }`.

This seems clearly like a bug to me and [@nikomatsakis thought so too](https://github.com/rust-lang/rust/issues/57593#issuecomment-454119940).

As usual with C-future-compatibility warnings, this will become a hard error eventually. In this case we should be able to do it fairly quickly given the sole regression found in crater.

@rfcbot merge

**Comment by Centril** [comment]
> @petrochenkov @Centril This is now a future incompatibility lint. Should we have an FCP or something?

Sure, let's do it...

I propose that we introduce this [**C-future-compatibility**](https://github.com/rust-lang/rust/labels/C-future-compatibility) warning ([*1 regression on crater*](https://github.com/rust-lang/rust/pull/57617#issuecomment-455341547)) that catches *multiple macro metavariables with the same name*, e.g. `macro_rules! b { ($a:expr, $a:expr) => {} }`.

This seems clearly like a bug to me and [@nikomatsakis thought so too](https://github.com/rust-lang/rust/issues/57593#issuecomment-454119940).

As usual with C-future-compatibility warnings, this will become a hard error eventually. In this case we should be able to do it fairly quickly given the sole regression found in crater.

@rfcbot merge

File: src/librustc_lint/lib.rs (hunk 4 above)
Comment (mark-i-m) [reply src/librustc_lint/lib.rs]: Created #57742

**Comment by mark-i-m** [comment]
@petrochenkov @Centril I believe I have addressed your comments.

**Comment by petrochenkov** [comment]
r=me once lang team review is complete.

**Comment by mark-i-m** [comment]
The nes crate has merged my PR, and also FWIW I think this lint found a real bug.

**Comment by Centril** [comment]
Review pings: @aturon @scottmcm @withoutboats, and @pnkfelix -- please have a look at this :)

**Comment by mark-i-m** [comment]
@Centril could you remove the "waiting-for-team" tag please?

**Comment by mark-i-m** [comment]
@Centril should this be mentioned in release notes?

**Comment by Centril** [comment]
@mark-i-m I think we usually do that when it becomes an actual hard error, but I'm not sure... cc @nikomatsakis re. that...

**Pull request merged**